flask/src/main/java/net/corda/gradle/flask/FlaskJarTask.java

```diff
@@ -0,0 +1,268 @@
+                try (InputStream is = new FileInputStream(fileCopyDetails.getFile())) {
+                    zoos.putNextEntry(zipEntry);
+                    Flask.write2Stream(is, zoos, buffer);
+                }
+            }
+        }
+    }
+
+    @Override
+    protected CopyAction createCopyAction() {
+        File destination = getArchiveFile().get().getAsFile();
+        return new CopyAction() {
+            @Override
+            @SneakyThrows
+            public WorkResult execute(CopyActionProcessingStream copyActionProcessingStream) {
+                Manifest manifest = new Manifest();
+                Attributes mainAttributes = manifest.getMainAttributes();
+                mainAttributes.put(Attributes.Name.MANIFEST_VERSION, "1.0");
+                mainAttributes.put(Attributes.Name.MAIN_CLASS, Flask.Constants.DEFAULT_LAUNCHER_NAME);
+                mainAttributes.putValue(Flask.ManifestAttributes.LAUNCHER_CLASS, launcherClassName.get());
+                mainAttributes.putValue(Flask.ManifestAttributes.PREMAIN_CLASS, Flask.Constants.DEFAULT_LAUNCHER_NAME);
+                Optional.ofNullable(mainClassName.getOrNull()).ifPresent(it ->
+                        mainAttributes.putValue(Flask.ManifestAttributes.APPLICATION_CLASS, it));
+                MessageDigest md = MessageDigest.getInstance("SHA-256");
+                mainAttributes.putValue(Flask.ManifestAttributes.HEARTBEAT_AGENT_HASH,
+                Flask.bytes2Hex(Flask.computeDigest(HeartbeatAgentResource.instance::read, md)));
+                byte[] buffer = new byte[Flask.Constants.BUFFER_SIZE];
+
+                /**
+                 * The manifest has to be the first zip entry in a jar archive, as an example,
+                 * {@link java.util.jar.JarInputStream} assumes the manifest is the first (or second at most)
+                 * entry in the jar and simply returns a null manifest if that is not the case.
+                 * In this case the manifest has to contain the hash of all the jar entries, so it cannot
+                 * be computed in advance, we write all the entries to a temporary zip archive while computing the manifest,
+                 * then we write the manifest to the final zip file as the first entry and, finally,
+                 * we copy all the other entries from the temporary archive.
+                 */
+                File temporaryJar = new File(getTemporaryDir(), "premature.zip");
+                try (ZipOutputStream zipOutputStream = new ZipOutputStream(new FileOutputStream(temporaryJar))) {
+                    StreamAction streamAction = new StreamAction(zipOutputStream, manifest, md, buffer);
+                    copyActionProcessingStream.process(streamAction);
+                }
+
+                try (ZipOutputStream zipOutputStream = new ZipOutputStream(new FileOutputStream(destination));
+                     ZipInputStream zipInputStream = new ZipInputStream(new FileInputStream(temporaryJar))) {
```

Comment:
I suspect all of these File streams should be buffered so that the Zip streams don't hammer the I/O. Perhaps even create some `getBufferedInputStream(File)`, `getBufferedOutputStream(File)` functions?